cookbook/core/control_flow/checkpoint.py

```diff
@@ -0,0 +1,76 @@
+provide exactly this capability. But, this is still less then ideal, as in the best case scenario, where failure is not
+common, the overhead of spawning new tasks, recording intermediates and re-bootstrapping the state can be extremely
+expensive. An example of this case is model-training. Running multiple epochs or different iterations with the same
+dataset can take a long time, but the bootstrap time may be high and creating task boundaries can be expensive.
+
+To tackle this, Flyte offers a way to checkpoint progress within a task execution as a file or a set of files. These
+checkpoints can be written synchronously or async and in case of failure, the checkpoint file can be re-read to resume
+most of the state without re-running the entire task. This opens up the opportunity to use alternate compute systems with
+lower guarantees like - `AWS Spot Instances <https://aws.amazon.com/ec2/spot/>`__, `GCP Pre-emptible Instances <https://cloud.google.com/compute/docs/instances/preemptible>`__ etc.
+
+These instances offer great performance at much lower price-points as compared to their OnDemand or Reserved alternatives.
+This is possible if you construct your tasks in a fault-tolerant way. For most cases, when the task runs for a short duration,
+less than 10 minutes, the potential of failure is not significant and Task-boundary based recovery offers great fault-tolerance
+to ensure successful completion.
+
+But, as the time for a task increases, the cost of re-running the task increases and the chance of it successfully completing reduces.
+This is where Flyte's intra-task checkpointing truly shines. This document provides an example of how to develop tasks which
+utilize intra-task checkpoining. This only provides the low-level API. We intend to integrate higher level api;s with popular
+training frameworks like Keras (ModelCheckpoint), Pytorch, Scikit learn, Spark (checkpoints), flink (checkpoints) to super charge
+their fault-tolerance.
+"""
+
+from flytekit import task, workflow, current_context
+from flytekit.exceptions.user import FlyteRecoverableException
+
+
+# %%
+# This task shows how in case of failure checkpoints can help resume. This is only an example task and shows the api for
+# the checkpointer. The checkpoint system exposes other api's and for a more detailed understanding refer to
+# :ref:pyclass:`flytekit.Checkpoint`
+#
+# The goal of this method is to actually return `a+4` It does this in 3 retries of the task, by recovering from previous
+# failures. For each failure it increments the value by 1
+@task(retries=3)
+def use_checkpoint(a: int) -> int:
+    cp = current_context().checkpoint
+    prev = cp.read()
+    v = a
+    if prev:
+        v = int(prev.decode())
+        if v - a > 2:
+            return v + 1
+    cp.write(f"{v + 1}".encode())
+    raise FlyteRecoverableException(f"V value {v + 1}")
```

Comment:
can we modify this example to somewhat emulate a model-training loop?
For example:
- include a for loop in the task
- write for loop iteration progress to the checkpoint
- reading the checkpoint should resume the loop from the latest checkpoint

```python
import random

@task(retries=3)
def use_checkpoint(n_iterations: int) -> int:
    cp = current_context().checkpoint
    prev = cp.read()

    start = 0
    if prev:
        start = int(prev.decode())

    for i in range(start, n_iterations):

        # simulate a random failure
        if random.random() > 0.8:
            raise FlyteRecoverableException(f"progress value {v + 1}")

        # save progress state
        cp.write(f"{i + 1}".encode())

    # in this simple example, just return the last value of the for loop
    return i
```

Reply:
One problem with this is, that there is no guarantee of failures (in random) and no guarantee of completion within the retries. 
I like the idea of training  loop, but I will make it a deterministic failure

Reply:
I changed it to this
```python
@task(retries=RETRIES)
def use_checkpoint(n_iterations: int) -> int:
    cp = current_context().checkpoint
    prev = cp.read()
    start = 0
    if prev:
        start = int(prev.decode())

    # We will create a failure interval so that we can create failures ever n iterations and then succeed within
    # configured retries
    failure_interval = n_iterations * 1.0 / RETRIES
    i = 0
    for i in range(start, n_iterations):
        # simulate a deterministic failure, for demonstration. We want to show how it eventually completes within
        # the given retries
        if i > start and i % failure_interval == 0:
            raise FlyteRecoverableException(f"Failed at iteration {start}, failure_interval {failure_interval}")
        # save progress state. It is also entirely possible save state every few intervals.
        cp.write(f"{i + 1}".encode())

    return i
```